### PR TITLE
Save scheduler execution time by caching dags

### PR DESCRIPTION
**Pull request by AutomationDev85**
Hi airflow community,
this is my second PR and be happy to work on the scheduler runtime again. We faced an issue with slow scheduler execution time by having millions of queued dag_runs for one DAG. 

This PR will add at 2 points in the code a caching of the dag. This saved a lot of scheduler runtime during scheduling many dag_runs for the same dag. The code currently reads the dag out of the DB and if you have a lot of short running tasks this is executed a lot. E.g. we wanted to schedule a DAG with:
max_active_tasks=60,
max_active_runs=180,
and most tasks with an execution time of 2 sec and 1 million in queued state.  With the caching we were able to increase scheduler performance. Because the time on our slow DB to query the dag took between 50ms and 250ms and if you execute this only once or 60 times during one scheduler loop run this makes a big change.

@vandonr-amz fyi, as discussed with @jens-scheffler-bosch

### CODE REVIEW

File: airflow/jobs/scheduler_job_runner.py
Comment (vandonr-amz) [inline airflow/jobs/scheduler_job_runner.py]: nit: when doing this, you do an extra lookup in the dict (when doing `x in dict.keys`) that can be avoided.

Here is how you can rewrite it to minimize dict accesses:
```
dag = cached_dags.get(dag_run.dag_id)  # returns None on miss
if dag is None:
    dag = self.dagbag.get_dag(dag_run.dag_id, session=session)
    cached_dags[dag_run.dag_id] = dag
```

Also, this code is duplicated between the two methods. It'd be nice to wrap this into a `get_or_insert(dict, key, lambda)` function that could go in some utils file maybe ?

Comment (AutomationDev85) [reply airflow/jobs/scheduler_job_runner.py]: Thanks for your feedback! This is a good idea, i added a function to the helpers, not sure is this is the right place but looking forward for your feedback :)

File: airflow/jobs/scheduler_job_runner.py
Comment (vandonr-amz) [inline airflow/jobs/scheduler_job_runner.py]: Passing the params individually is a convoluted way to do it... Using a lambda you can just pass the whole packaged call as a Callable without the need for extra parameters
```suggestion
            dag = get_value_with_cache(
                cached_dags, dag_run.dag_id, lambda: self.dagbag.get_dag(dag_run.dag_id, session=session)
            )
```

File: airflow/jobs/scheduler_job_runner.py
Comment (vandonr-amz) [inline airflow/jobs/scheduler_job_runner.py]:
```suggestion
            dag = dag_run.dag = get_value_with_cache(
                cached_dags, dag_run.dag_id, lambda: self.dagbag.get_dag(dag_run.dag_id, session=session)
            )
```

File: airflow/utils/helpers.py
Comment (vandonr-amz) [inline airflow/utils/helpers.py]: By using a lambda, we don't need to handle paramters here. (suggesting a rename of the function passed as well ? WDYT ?)

Also, `if not return_value` will trigger if the saved value is "falsy", which is the case for None, but also for empty lists/dict for instance, which can lead to surprising behavior. I think explicitly checking against None would be better here.

```suggestion
def get_value_with_cache(cache: dict, key: str, do_get_value: Callable) -> Any:
    """Returns value from cache or function"""
    return_value = cache.get(key)
    if return_value is None:
        return_value = cache[key] = do_get_value()

    return return_value
```

Comment (vandonr-amz) [reply airflow/utils/helpers.py]: I'll let you fix the test if you apply this change ;)

File: airflow/jobs/scheduler_job_runner.py
Comment (AutomationDev85) [reply airflow/jobs/scheduler_job_runner.py]: Thanks for the feedback! Using now lambda.

File: airflow/jobs/scheduler_job_runner.py
Comment (AutomationDev85) [reply airflow/jobs/scheduler_job_runner.py]: Thanks for the feedback! Using now lambda.

File: airflow/utils/helpers.py
Comment (AutomationDev85) [reply airflow/utils/helpers.py]: Thanks for the feedback. I like the new name for the function! 
I'm using `if not return_value` because if the cache returns empty value, I think the best way is to call do_get_value to make sure that the right and latest value is used and not the cached value.

Comment (vandonr-amz) [reply airflow/utils/helpers.py]: ok makes sense then. That specific case doesn't really matter for the current application anyway.

**Review by vandonr-amz — APPROVED** [review]
me gusta

**Review by potiuk — APPROVED** [review]
Looks like nice optimisation..  I cannot see any bad side-effects of it.

But I'd love another maintainer's opinion.

**Comment by uranusjr** [comment]
The added function can basically be replaced by `functools.lru_cache`.

**Comment by potiuk** [comment]
> > This functionality is basically `functools.lru_cache`.
> 
> yes and no because this is an in-function cache. `functools.lru_cache` is intended to provide a cache that is persisted at the object level I think ?

Not realy. @uranusjr is right. lru_cache caches the value of function keyed by arguments  of the function (the only condition is that the arguments must be hashable). So it does precisely what the extra (unneded) code does in this case..

**Comment by vandonr-amz** [comment]
> lru_cache caches the value of function keyed by arguments of the function (the only condition is that the arguments must be hashable). So it does precisely what the extra (unneded) code does in this case..

ok, but where is that value cached ?
If the annotated method is global, the cache is global I suppose ? If it's an class method, the cache is stored with the object instance ?
Here we want to cache just within the scope of the function, so we could write an annotated method as a sub-method, but then we have to write it twice, in both locations where it's used ?

We can certainly do that, but I suggested extracting the duplicated code to a function in the first place to avoid this ^^

Or maybe there is a third way I'm missing ? With an lru_cache that is not duplicated and that doesn't cache _too much_ ?

**Comment by jens-scheffler-bosch** [comment]
> > > This functionality is basically `functools.lru_cache`.
> > 
> > 
> > yes and no because this is an in-function cache. `functools.lru_cache` is intended to provide a cache that is persisted at the object level I think ?
> 
> Not realy. @uranusjr is right. lru_cache caches the value of function keyed by arguments of the function (the only condition is that the arguments must be hashable). So it does precisely what the extra (unneded) code does in this case..

Only being a 95% Python expert (need to earn some stars), removing code with functools sounds good. But also I suppose adding a `@lru_cache` to `self.dagbag.get_dag()` on a general level is not what we want - so to only locally optimize I assume the best approach would be also using a `lambda` like proposed in answer #1 here? https://stackoverflow.com/questions/10270360/python-use-lru-cache-on-lambda-function-or-other-ways-to-create-cache-for-lamb

**Comment by vandonr-amz** [comment]
oh nice, I didn't know it could be used like that ! Looks great to me !

**Review by AutomationDev85 — COMMENTED** [review]
Nice idea to use lru_cache. Thanks for helping newbie to cross the street.

**Comment by potiuk** [comment]
> Or maybe there is a third way I'm missing ? With an lru_cache that is not duplicated and that doesn't cache too much ?

1. Make a class methods wrapping the global method call
2. Annotate it with @lru_cache

Saves about 80% of the code.

**Comment by potiuk** [comment]
BTW. If you really want, you could even annotate it within the non-anonymous lambda (ie. function inside function. Function in Python can easily be declared and run inside another function. 


```

class Class:

      def method(self, .....):
      
          @lru_cache()...
           def cached_call():
               # body of cached call
               ....
          
          # body of object method 
          ....

```

**Comment by vandonr-amz** [comment]
I like the lambda solution, I think it looks cleaner than having a function inside the function

**Comment by potiuk** [comment]
> I like the lambda solution, I think it looks cleaner than having a function inside the function

Yes it has an appeal. the most important is that we are building of lru_cache, rather than trying to do what it does

**Comment by potiuk** [comment]
Second pair of eyes needed here , as this is core part.

**Comment by potiuk** [comment]
Second pair of 👀 and 🙌  needed :)

File: airflow/jobs/scheduler_job_runner.py
Comment (uranusjr) [inline airflow/jobs/scheduler_job_runner.py]: Could benefit from using `functools.partial` instead. Also it’d be a good idea to improve the type hint here so when the callable is called its return value can also be typed instead of being a useless `Any`.

Comment (jscheffl) [reply airflow/jobs/scheduler_job_runner.py]: I tried to think about how the `functools.partial` could make it better. The proposed statement is a one-liner in adding the `@lru_cache` decorator inline w/o a local additional function declaration. How do you mean it would be better?

In my VSCode at least it shows type hints from parsing python, resulting callable is hinted and the returned `dag` variable in line 1090 is displayed as `(variable) dag: DAG | None` - is it displayed w/o type hints in your IDE?

Comment (uranusjr) [reply airflow/jobs/scheduler_job_runner.py]: VS Code uses Pylance, which uses additional guessing to infer types. The heuristics mostly work well and are very helpful, but are not always flawless. Correctly annotating the callable would be more friendly for people using other tools.

Mypy identifies the type to be `Any` (using `typing.reveal_type()` to inspect):

```
$ mypy airflow/jobs/scheduler_job_runner.py 
airflow/jobs/scheduler_job_runner.py:1092: note: Revealed type is "Any"
```

Comment (jscheffl) [reply airflow/jobs/scheduler_job_runner.py]: Thanks for the proposal and thinking a moment about this, it makes the code really "nicer" now. As @AutomationDev85 is off on vacation I'm pushing the update "on behalf".

**Review by uranusjr — APPROVED** [review]

**Comment by uranusjr** [comment]
Pending CI

**Comment by ashb** [comment]
@uranusjr @jens-scheffler-bosch @potiuk This PR makes no sense btw -- DagBag.get_dag _already_ has a cache:

https://github.com/apache/airflow/blob/d1c58d86de1267d9268a1efe0a0c102633c051a1/airflow/models/dagbag.py#L189-L192

**Pull request merged**
